src/logoutd.c

```diff
@@ -44,11 +46,7 @@ static int check_login (const struct utmp *ut)
 	char user[sizeof (ut->ut_user) + 1];
 	time_t now;
 
-	/*
-	 * ut_user may not have the terminating NUL.
-	 */
-	strncpy (user, ut->ut_user, sizeof (ut->ut_user));
-	user[sizeof (ut->ut_user)] = '\0';
+	zustr2stp(user, ut->ut_user, SIZEOF_ARRAY(ut->ut_user));
```

Comment:
string_copying(7) defines zustr as a null terminated character sequence.  Should a ustr2stp be used here instead?

In either case, I don't think the 'ut_user may not have the terminating NUL' comment should be removed.  You may think it should be obvious, but I don't believe it is.

Reply:
I also disagree with the claim that the length of dest buffer does not matter.  In fact, there's little more important than avoiding overwriting the end of the dest buffer.

In the zustr2stp, I would expect the source to be a zustr, zero terminated, so source length is dictated by the ending null and shouldn't need to be listed.

In a ustr2stp, I would think both lengths would need to be given.  The source length since it may not have an ending null, and dest length because you must not overwrite it.

We are "handling" it in these cases by allocating enough space in the dest.  However, there's nothing tying the alloc length and source-buffer-lengths explicitly, so after some code churn while noone's noticing, the relation could be broken, and user could be defined as being length UT_NAMESIZE/2.

Reply:
> string_copying(7) defines zustr as a null terminated character sequence. 

No.  Read carefully; it says "null-padded", not null-terminated.

With null-padded I mean that if there's remaining space in the array after the character sequence, those bytes are zero.  But there may be no remaining space, so there's no guarantee of a termination.

If it were null-terminated, it would be a normal string (`str`).  The mnemonics for `zustr` are `u` for unterminated, and `z` for zero-padded (which implies a fixed-width array).

> Should a ustr2stp be used here instead?

Nope.

> 
> In either case, I don't think the 'ut_user may not have the terminating NUL' comment should be removed.

The use of `zustr` implies that the source is not necessarily NUL-terminated.  `u` in `zustr` means that.

> You may think it should be obvious, but I don't believe it is.

If you still believe it, I can reconsider, but please reconsider.  :)

Reply:
If you want to be a bit paranoic, this code can help:

```diff
diff --git a/lib/zustr2stp.h b/lib/zustr2stp.h
index 7775fc6d..6bec0390 100644
--- a/lib/zustr2stp.h
+++ b/lib/zustr2stp.h
@@ -10,6 +10,7 @@
 
 #include <config.h>
 
+#include <assert.h>
 #include <stddef.h>
 #include <string.h>
 
@@ -17,7 +18,15 @@
 #include "sizeof.h"
 
 
-#define ZUSTR2STP(dst, src)  zustr2stp(dst, src, SIZEOF_ARRAY(src))
+#define ZUSTR2STP(dst, src)                                                   \
+({                                                                            \
+       size_t  ssize;                                                        \
+                                                                              \
+       ssize = SIZEOF_ARRAY(src);                                            \
+       static_assert(!is_array(dst) || sizeof(dst) > ssize, "");             \
+                                                                              \
+       zustr2stp(dst, src, SIZEOF_ARRAY(src);                                \
+})
 
 
 inline char *zustr2stp(char *restrict dst, const char *restrict src, size_t sz);
```

However, I don't expect it should catch anything more than `_FORTIFY_SOURCE` would already do.

Reply:
> With null-padded I mean that if there's remaining space in the array after the character sequence, those bytes are zero. But there may be no remaining space, so there's no guarantee of a termination.

If I can make a man-pages request, I think it would be best to state that explicitly there when defining the terms.

Reply:
Sure you can!  I will check that page with this in mind.

Reply:
Regarding the added assert - I'm ok not doing that, so long as we add a FORTIFY_SOURCE=2 build to CI :)  And then instead of the assert, a comment at the top explaining that that's what we're relying on to catch newly introduced errors (since myself and others may wonder again in a few months).

I'm also ok with just adding the static_assert.

Reply:
I'll add the static_assert(3).  It's a nice reminder of assumptions.  More or less the same as a comment, but checked by the compiler.  :)

Reply:
Hmm, I'm finding difficulties.  I would have expected this to be a compile-time constant, but it seems it isn't!

```c
In file included from logoutd.c:23:
logoutd.c: In function ‘check_login’:
../lib/zustr2stp.h:27:38: error: expression in static assertion is not constant
   27 |         static_assert(!is_array(dst) || sizeof(dst) > ssize_, "");            \
      |                       ~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~
logoutd.c:49:9: note: in expansion of macro ‘ZUSTR2STP’
   49 |         ZUSTR2STP(user, ut->ut_user);
      |         ^~~~~~~~~
logoutd.c: In function ‘main’:
../lib/zustr2stp.h:27:38: error: expression in static assertion is not constant
   27 |         static_assert(!is_array(dst) || sizeof(dst) > ssize_, "");            \
      |                       ~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~
logoutd.c:227:25: note: in expansion of macro ‘ZUSTR2STP’
  227 |                         ZUSTR2STP(user, ut->ut_user);
      |                         ^~~~~~~~~
```

Reply:
D'oh!  It was my fault for using a temporary variable.  Me bad.  :)

Reply:
Done.